packages/drivers/odsp-driver/src/odspSummaryUploadManager.ts

```diff
@@ -0,0 +1,323 @@
+        pathToBlobSha: new Map(),
+        treesPathToTree: new Map(),
+    };
+
+    // Last proposed handle of the uploaded app summary.
+    private lastSummaryProposalHandle: string | undefined;
+
+    constructor(
+        private readonly snapshotUrl: string,
+        private readonly getStorageToken: (options: TokenFetchOptions, name?: string) => Promise<string | null>,
+        private readonly logger: ITelemetryLogger,
+        private readonly epochTracker: EpochTracker,
+        private readonly blobCache: Map<string, IBlob | ArrayBuffer> = new Map(),
+    ) {
+    }
+
+    public async buildCachesForDedup(snapshotTree: api.ISnapshotTree, path: string = ""): Promise<api.ISummaryTree> {
+        assert(Object.keys(snapshotTree.commits).length === 0, "There should not be commit tree entries in snapshot");
+
+        const summaryTree: api.ISummaryTree = {
+            type: api.SummaryType.Tree,
+            tree: {},
+        };
+        for (const [key, value] of Object.entries(snapshotTree.blobs)) {
+            const blobValue = this.blobCache.get(value);
+            assert(blobValue !== undefined, "Blob should exists");
+            const decodedBlobValue = {
+                content: blobValue instanceof ArrayBuffer ? IsoBuffer.from(blobValue).toString("utf8") : fromBase64ToUtf8(blobValue.content),
+                encoding: "utf-8",
+            };
+            const hash = await hashFile(IsoBuffer.from(decodedBlobValue.content, decodedBlobValue.encoding));
+            // We are just setting the hash as content but we won't use it anywhere as it doesn't represent the actual content of blob.
+            summaryTree.tree[key] = {
+                type: api.SummaryType.Blob,
+                content: hash,
+            };
+            const fullBlobPath = path === "" ? `.app/${key}` : `${path}/${key}`;
+            this.blobTreeDedupCaches.blobShaToPath.set(hash, fullBlobPath);
+            this.blobTreeDedupCaches.pathToBlobSha.set(fullBlobPath, hash);
+        }
+
+        for (const [key, tree] of Object.entries(snapshotTree.trees)) {
+            const fullTreePath = path === "" ? key === ".protocol" ? `${key}` : ".app" : `${path}/${key}`;
+            const subtree = await this.buildCachesForDedup(tree, fullTreePath);
+            this.blobTreeDedupCaches.treesPathToTree.set(fullTreePath, subtree);
+            summaryTree.tree[key] = subtree;
+        }
+        return summaryTree;
+    }
+
+    public async writeSummaryTree(tree: api.ISummaryTree, context: ISummaryContext) {
+        // If the last proposed handle is not the proposed handle of the acked summary(could happen when the last summary get nacked),
+        // then re-initialize the caches with the backup ones.
+        if (context.proposalHandle !== this.lastSummaryProposalHandle) {
+            this.blobTreeDedupCaches = { ...this.backupBlobTreeDedupCaches };
+        }
+        this.backupBlobTreeDedupCaches = { ...this.blobTreeDedupCaches };
```

Comment:
Maybe name it to reflect that it tracks "last acked summary status"?
You can put this line into else block

Reply:
Added it in else part. But the name "lastSummaryProposalHandle' is correcct as it is the last summary proposal handle. We don't know whether it got acked or not. That is why we have a previous cache too.